### ISSUE TITLE
Loss of accuracy when GPT-2 based model is exported to ONNX

### ISSUE THREAD

**Issue by gurunathparasaram** [microsoft/onnxruntime#6549]
**Performance loss**
 - My distilgpt2 model exported from huggingface to ONNX when run in ONNX-Runtime doesn't yield desired accuracy.
 

**System information**
- OS Platform and Distribution: Linux Ubuntu 20.04
- ONNX Runtime installed from: PyPI 
- ONNX Runtime version: 1.5.2
- Python version: 3.7.9
- GCC/Compiler version (if compiling from source): Not applicable
- I tried on CPU, so following are not applicable.
- Driver Version: 460.32.03   
- CUDA Version: 11.2
- GPU model and memory: Quadro RTX 8000

**Details:**
- I have been trying out GPT-2 model for sentence-scoring for correcting sentences which have homophones.
__Example:__ To ensure correct usage of homophones in sentences, I used huggingface's distilgpt2 to see which of ["The bare is in the forest", "The bear is in the forest."] is more probable.
 - For this use-case, I get the outputs from the distilgpt2 model, and take cross-entropy loss between inputs & labels and then consider the sentence from a set of candidate sentences more probable if it has the least loss.
 - While the original huggingface model gives precision, recall and F-0.5 score in the range of 0.90-0.95, the same model when exported to ONNX as shown in [this example notebook](https://github.com/microsoft/onnxruntime/blob/1ce2982f65e5516067fdcaef19409279173b0d75/onnxruntime/python/tools/transformers/notebooks/Inference_GPT2_with_OnnxRuntime_on_CPU.ipynb) gives lower precision, recall and F-0.5 scores. F-score drops to around 0.82-0.83. Also, there is about 30% drop in true positives.
 - On further observation, I found that scores between the unconverted HF model and the converted model sometimes vary in the 4th decimal place.
 - Also, observed a similar performance drop earlier when we converted  and used a TokenClassification model from huggingface in ONNX-Runtime.
 - Is this variation the likely cause of this performance drop while scoring sentences with homophones? What can be done to ensure that performance drop is as small as possible? Please mention here if any settings, etc., need to be changed / configured or if I am missing something. Any pointers to resources would be helpful. Thanks.

CPU flags if this helps:
```Model:                           85
Model name:                      Intel(R) Xeon(R) Gold 5218 CPU @ 2.30GHz
Stepping:                        7
CPU MHz:                         1040.127
CPU max MHz:                     3900.0000
CPU min MHz:                     1000.0000
BogoMIPS:                        4600.00
Virtualization:                  VT-x
L1d cache:                       1 MiB
L1i cache:                       1 MiB
L2 cache:                        32 MiB
L3 cache:                        44 MiB
NUMA node0 CPU(s):               0-15,32-47
NUMA node1 CPU(s):               16-31,48-63
Vulnerability Itlb multihit:     KVM: Mitigation: VMX disabled
Vulnerability L1tf:              Not affected
Vulnerability Mds:               Not affected
Vulnerability Meltdown:          Not affected
Vulnerability Spec store bypass: Mitigation; Speculative Store Bypass disabled via prctl and seccomp
Vulnerability Spectre v1:        Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:        Mitigation; Enhanced IBRS, IBPB conditional, RSB filling
Vulnerability Srbds:             Not affected
Vulnerability Tsx async abort:   Mitigation; TSX disabled
Flags:                           fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts acpi mmx fxsr sse sse2 ss ht tm pbe syscall nx pdpe1gb rdtscp lm constant_tsc art arch_perfmon pebs bts rep_good nopl xtopology nonstop_tsc cpuid aperfmperf pni pclmulqdq dtes64 monitor ds_cpl vmx smx est tm2 ssse3 sdbg fma cx16 xtpr pdcm pcid dca sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand lahf_lm abm 3dnowprefetch cpuid_fault epb cat_l3 cdp_l3 invpcid_single intel_ppin ssbd mba ibrs ibpb stibp ibrs_enhanced tpr_shadow vnmi flexpriority ept vpid ept_ad fsgsbase tsc_adjust bmi1 avx2 smep bmi2 erms invpcid cqm mpx rdt_a avx512f avx512dq rdseed adx smap clflushopt clwb intel_pt avx512cd avx512bw avx512vl xsaveopt xsavec xgetbv1 xsaves cqm_llc cqm_occup_llc cqm_mbm_total cqm_mbm_local dtherm ida arat pln pts pku ospke avx512_vnni md_clear flush_l1d arch_capabilities
```

**Comment by yuslepukhin** [comment]
Just to clarify, is performance an issue or it is just an accuracy?

**Comment by gurunath-p** [comment]
It is an accuracy issue. Sorry for not being clear enough earlier. I should have phrased it better. Thanks.

**Comment by faxu** [comment]
CC @tianleiwu

**Comment by tianleiwu** [comment]
For float32 models, GPT-2 usually generate same text (>99%) as PyTorch result using greedy method. 

If you use float16 or int8 model, it is expected that there will be accuracy loss, and generated text is likely be quite different.

Please share a script that could reproduce the problem.

**Comment by sai-prasanna** [comment]
@tianleiwu Is the difference in output between onnx and pytorch to be expected? I see a difference in 3rd decimal place on wards.
 I think the problem here is that for the outputs even 3rd decimal place difference impacts accuracy.

**Comment by collinarnett** [comment]
I wanted to add my 2 cents here since this issue seems to be the same that I'm facing.
I have a finetuned GPT2-XL model from hf and I've quantized it using the workaround described in https://github.com/microsoft/onnxruntime/issues/7017 

This yeilds the following results when I do the logit diff in the https://github.com/microsoft/onnxruntime/blob/master/onnxruntime/python/tools/transformers/notebooks/Inference_GPT2_with_OnnxRuntime_on_CPU.ipynb on my model and the non-finetuned version of GPT2-XL

My Model:
float32 model onnx vs pytorch:
```
max logits diff (ignored padding) 2.86102294921875e-05
```
Quantized int8 onnx vs pytorch
```
max logits diff (ignored padding) 1.972315788269043
```

Base GPT-XL:
float32 model onnx vs pytorch:
```
max logits diff (ignored padding) 3.62396240234375e-05
```
Quantized int8 onnx vs pytorch:
```
max logits diff (ignored padding) 2.5597681999206543
```

The thing is this doesn't happen with the smaller models in the notebook. Which makes me think that the `infer_shapes` function has something to do with it as mentioned in the other issue I linked.

**Comment by isgursoy** [comment]
Observing biger difference in exactly same use case. fp32, distilgpt2.

**Comment by tianleiwu** [comment]
For anyone interested in GPT-2 parity, we have a tool:
https://github.com/microsoft/onnxruntime/blob/master/onnxruntime/python/tools/transformers/gpt2_parity.py
It test the accuracy of generating one word and compare the result with PyTorch

Example to run:
   python -m onnxruntime.transformers.gpt2_parity -m gpt2-large --use_gpu -e

For GPT-2 model, we see 100% matched with PyTorch.  Mixed precision (Add, FastGelu and LayerNormalizaiton comptued in FP32, and other operators in FP16), top 1 accuracy of one word is 99.1~99.5% depending on the model, and the accuracy of whole sentence will drop with generating more words.

int8 quantization are not recommended due to accuracy loss (shall be larger than mixed precision).
